### PR TITLE
fix: survey preview for suid enabled surveys

### LINKED ISSUE #6244: Survey preview not working for SUID-enabled link surveys
### Issue Summary

When I click on “Preview” from the link survey summary page, I’m redirected to a “survey link invalid” screen.

Image

### Expected Behavior

The survey preview should render as expected.

To solve the issue our preferred solution is that if single use ids are used, the backend generates a valid (always encrypted) single use id that is also appended to the preview url.

### Other information (incl. screenshots, Formbricks version, steps to reproduce,...)

**This is a critial issue blocking the `v3.16.0` QA and after merged to main also needs to be backported to the `release/v3.16` branch.**

Steps to reproduce:
1. Create a link survey.
2. Go to the **editor settings > Response Options**.
3. Toggle the "Single Use" survey link.
4. Try to open the survey preview from the summary page.

### Your Environment

None

### PR DESCRIPTION
**Pull request by gupta-piyush19**
## What does this PR do?
survey preview for suid enabled surveys

Fixes #6244 

## How should this be tested?

1. Create a link survey.
2. Go to the editor settings > Response Options.
3. Toggle the "Single Use" survey link.
4. Try to open the survey preview from the summary page.

## Summary by CodeRabbit

* **New Features**
  * Survey preview links now support single-use IDs, enhancing security for surveys with single-use enabled.
  * Preview URLs are generated asynchronously to include single-use IDs when applicable.

* **Bug Fixes**
  * Improved error handling and user feedback during survey duplication and preview actions.

* **Tests**
  * Expanded test coverage for survey duplication, preview, sharing, status management, and edge cases.
  * Enhanced async handling and mocking in test suites for more reliable results.

### CODE REVIEW

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/components/SurveyAnalysisCTA.tsx [hunk 1]
```diff
@@ -102,8 +106,17 @@ export const SurveyAnalysisCTA = ({
     setLoading(false);
   };
 
-  const getPreviewUrl = () => {
+  const getPreviewUrl = async () => {
     const separator = surveyUrl.includes("?") ? "&" : "?";
```
Comment (pandeymangg) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/components/SurveyAnalysisCTA.tsx]: Can we replace the separator logic here with actually manipulating URLs? then we can simply do something like `surveyUrl.searchParams.set("suid", newId)`.

File: apps/web/modules/analysis/components/ShareSurveyLink/index.tsx [hunk 2]
```diff
@@ -32,6 +33,21 @@ export const ShareSurveyLink = ({
     setSurveyUrl(url);
   };
 
+  const { refreshSingleUseId } = useSingleUseId(survey);
+
+  const getPreviewUrl = async () => {
+    const separator = surveyUrl.includes("?") ? "&" : "?";
```
Comment (pandeymangg) [inline apps/web/modules/analysis/components/ShareSurveyLink/index.tsx]: same thing here as well

**Review by pandeymangg — CHANGES_REQUESTED** [review]
Thanks for the PR! @gupta-piyush19 It works great!
I just left a small comment, pls take a look 🙏

**Review by pandeymangg — APPROVED** [review]
LGTM

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/components/SurveyAnalysisCTA.tsx [hunk 3]
```diff
@@ -102,9 +102,18 @@ export const SurveyAnalysisCTA = ({
...
-    return `${surveyUrl}${separator}preview=true`;
+  const getPreviewUrl = async () => {
+    const surveyUrl = new URL(`${publicDomain}/s/${survey.id}`);
+
+    if (survey.singleUse?.enabled) {
+      const newId = await refreshSingleUseId({ encrypted: true });
+      if (newId) {
+        surveyUrl.searchParams.set("suId", newId);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (gupta-piyush19) [reply apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/components/SurveyAnalysisCTA.tsx]: This is fine, we aren’t throwing any errors here.

**Review by pandeymangg — APPROVED** [review]

**Review by mattinannt — APPROVED** [review]
@gupta-piyush19 Thanks for the PR and the recent changes.
Works great and once we made sure that the language changes are safe to merge we can merge the PR :-)
